diff --git a/script/buildtim.sh b/script/buildtim.sh
index 9c9dded..4f5bc79 100755
--- a/script/buildtim.sh
+++ b/script/buildtim.sh
@@ -9,12 +9,11 @@
 # $5 - Clocks preset
 # $6 - DDR topology map
 # $7 - Partition number
-# $8 - Build primary or secondary image (1/0)
-# $9 - Enable debug message (1/0)
-# $10 - Output TIM/NTIM file name
-# $11 - Output TIMN file name (valid for trusted mode only)
-# $12 - TIMN CSK sign key file name (valid for trusted mode only)
-# $13 - Encrypt the image (valid for trusted mode only)
+# $8 - Enable debug message (1/0)
+# $9 - Output TIM/NTIM file name
+# $10 - Output TIMN file name (valid for trusted mode only)
+# $11 - TIMN CSK sign key file name (valid for trusted mode only)
+# $12 - Encrypt the image (valid for trusted mode only)
 #
 
 DATE=`date +%d%m%Y`
@@ -24,9 +23,8 @@ WTPPATH=$4
 PRESET=$5
 DDRTOPOLOGY=$6
 BOOTPART=$7
-PRIMARY=$8
-DEBUG=$9
-OUTFILE=${10}
+DEBUG=$8
+OUTFILE=${9}
 
 CLOCKSPATH=$WTPPATH/tim/ddr
 SYSINITPATH=$WTPPATH/wtmi/sys_init
@@ -36,9 +34,6 @@ SYSINITLOADADDR=`sed -n '/. = 0x1fff[0-9|a-f]*/p' $SYSINITPATH/sys_init.ld | awk
 FILEEXT="txt"
 # Image definition files name prefix
 IMGPREF="img"
-if [ $PRIMARY == "0" ]; then
-	IMGPREF="secondary/img"
-fi
 # Number of images excepting TIM
 IMGSNUM=`ls -l $IMGPATH/$IMGPREF*.$FILEEXT | wc -l`
 # Reserved area definition for TIM/NTIM - file name prefix
