diff --git a/a3700/mv_ddr_tim.h b/a3700/mv_ddr_tim.h
index 38eae76..3dec06f 100644
--- a/a3700/mv_ddr_tim.h
+++ b/a3700/mv_ddr_tim.h
@@ -354,7 +354,7 @@ static char ddr4_1cs_tim_post[] = {
 	"DELAY: 0x0000000A\n"
 	"\n"
 	";Step9: DDRPHY Driver/Receiver & DQS internal Pullup/Pulldown settings\n"
-	"WRITE: 0xC0001004 0xD0677449\n"
+	"WRITE: 0xC0001004 0xD0133449\n"
 	"WRITE: 0xC0001008 0xC770055A\n"
 	"WRITE: 0xC000100C 0x5461DF77\n"
 	"\n"
@@ -363,7 +363,7 @@ static char ddr4_1cs_tim_post[] = {
 	";Step11: DDRPHY pads POWERDOWN disable\n"
 	"WRITE: 0xC0001030 0x03800000\n"
 	"WRITE: 0xC0001034 0x00000000\n"
-	"WRITE: 0xC0001090 0x00000004\n"
+	"WRITE: 0xC0001090 0x00000000\n"
 	"WRITE: 0xC0001094 0x00000000\n"
 	"\n"
 	";Step12: initC2Dofset DDR4\n"
